### ISSUE TITLE
Implement public partial sort Array API

### ISSUE THREAD

**Issue by iSazonov** [dotnet/corefx#25945]
## Rationale
- In [PowerShell repo](https://github.com/PowerShell/PowerShell) we have Sort-Object cmdlet enhanced with `-Top` and `-Bottom` parameters  - their implementation is based on partial sorting.
 - Currently Linq uses [`Array.Sort`](https://github.com/dotnet/corefx/pull/24116) but own internal (not public API) implementation of [PartialQuickSort](https://github.com/stephentoub/corefx/blob/a6aff797a33e606a60ec0c9ca034a161c609620f/src/System.Linq/src/System/Linq/OrderedEnumerable.cs#L592).
- It seems this is a very popular scenario for web interface to show sorted result by pages. Users use only top items and rarely go to the following pages.

## Proposed API ##
```c#
static void SortPartial(array array, int startIndex, int endIndex)
static void SortPartial(array keys, array items, int startIndex, int endIndex)
static void SortPartial(array array, int startIndex, int endIndex, System.Collections.IComparer comparer)
static void SortPartial(array keys, array items, int startIndex, int endIndex, System.Collections.IComparer comparer)
static void SortPartial[T](T[] array, int startIndex, int endIndex)
static void SortPartial[TKey, TValue](TKey[] keys, TValue[] items, int startIndex, int endIndex)
static void SortPartial[T](T[] array, int startIndex, int endIndex, System.Collections.Generic.IComparer[T] comparer)
static void SortPartial[TKey, TValue](TKey[] keys, TValue[] items, int startIndex, int endIndex, System.Collections.Generic.IComparer [TKey] comparer)

```

**Comment by iSazonov** [comment]
The initial implementation could be the use of `PartialQuickSort` from Linq. 
Ideally I guess it would enhance `private void IntrospectiveSort` from `Array`.

**Comment by JonHanna** [comment]
> It seems this is a very popular scenario for web interface to show sorted result by pages.

That's certainly where I got the idea of doing that partial quicksort from.

I'd prefer there was a yay or a nay on #25240 before exposing a partial sort, since they yay or nay should carry through to anything like this.

An in-place sort could *perhaps* benefit from setting the range to sort within as well as the range to get sorted, especially if using it incrementally. Say I have 100 items and I want the get 10 according to some order. If I then want the second "page" I ideally don't want to sort to seen items 10-19 from items 0-99, but items 10-19 from items 10-99; knowing that items 0-9 are already "below" the first item I'd want. This also stops the algorithm from moving things around in the already-sorted area and then giving up before they are back where they should be (since the whole gain of partial quick sort is that it gives up on chunks it knows it doesn't need to concern itself about). Note that `Sort` already allows one to set the range the sort happens within.

In terms of down-sides the only negative (as opposed to the all-features-start-out-unimplemented default negative) I can think of is the potential for confusion of already-sorted areas being unsorted by later partial sorts. We'd want to either rule that out in the implementation or make sure it was documented. Again, being able to set the range the sort happens within would mitigate this.

I think I prefer `length` to `endIndex`, it avoids confusion about whether `endIndex` is inclusive or exclusive (there are programming backgrounds that would strongly expect inclusive and programming backgrounds that would strongly expect exclusive, so each is strange to a some group of potential users) and is more inline with many `Array` methods such as `Copy`, `BinarySearch` and indeed, `Sort`.

**Comment by iSazonov** [comment]
@JonHanna Thanks for your feedback!
I think honest paging can only be for immutable arrays. So I don't know full right solution. My initial thoughts was about PowerShell scenario to get only N-top (-bottom) elements and web scenario to get first (ex.) 60 elements and show it by 20 elements (expecting 3-sigma rule - nobody use 4th page in Google).
I agree with `endIndex` to `length`.

**Comment by jkotas** [comment]
All new APIs like these should work on Spans, not arrays.

**Comment by karelz** [comment]
Do I understand it correctly, that the ask is to be able to sort just a specific segment of array?
In that case, adding `Sort` on `Span` would be the right thing to do as @jkotas pointed out.

**Comment by iSazonov** [comment]
No, partial sort does sort _entire array_ but return requested top/bottom elements only - the performance gain is that there is no need to sort the tail.

**Comment by karelz** [comment]
Ah, ok, that's what I originally thought, but the arg names `startIndex` and `endIndex` lead me to believe otherwise. Maybe we could use better names?
What is the `keys`/`items` for? Sorting dictionary?

**Comment by iSazonov** [comment]
`startIndex` and `endIndex` came from [PartialQuickSort](https://github.com/stephentoub/corefx/blob/a6aff797a33e606a60ec0c9ca034a161c609620f/src/System.Linq/src/System/Linq/OrderedEnumerable.cs#L592). Partial sort is a generalization of sorting and perhaps these names make sense. Although as suggested above we could use `index` and `length`.
The `keys/items` is from Array.Sort API. I don't know about dictionary sorting. We can see sample in [docs](https://msdn.microsoft.com/en-us/library/0h22cx4f(v=vs.110).aspx)

**Comment by iSazonov** [comment]
I think @jkotas meant https://github.com/dotnet/corefx/issues/15329. It is another proposal and it is currently being implemented by @nietras. 
If the proposal were approved, we could ask him to implement it also.

**Comment by nietras** [comment]
I am indeed currently implementing sorting for spans. As I understand this proposal, it sounds like some kind of partioning like `std::nth_element` i.e. partition so 0-9 are less than elements in 10-99, followed by a full sort on elements 0-9. One can the continue by partitioning 10-19 and 20-99, sort 10-19 etc.

Thus, this could be implemented by adding an equivalent function to `std:nth_element`. Not sure what a proper name for that would be in .NET `Partition` seems apt. I could try to make a proposal for that once I am done with sorting, if there is interest in this.

**Comment by JonHanna** [comment]
@nietras `nth_element` is a partitioning selection operation and hence uses a partitioning selection algorithm (generally introselect to avoid the worst-case of quickselect) there are partial sort algorithms that more directly do that operation.

**Comment by nietras** [comment]
> nth_element is a partitioning selection operation and hence uses a partitioning selection algorithm ... there are partial sort algorithms that more directly do that operation.

@JonHanna you are right, of course, it wasn't the best example since it is not exactly what I was proposing either. It isn't `std::partition` either so perhaps the name `Partition` is not ideal either... instead it is in fact `std::partial_sort` (not sure why I forgot that 😦) that could be used as a blueprint for similar functionality in C# i.e. something like:

```csharp
public static class MemoryExtensions
{
    public static void PartialSort<T>(this Span<T> items, int middle) 
        where T : IComparable<T>;
    public static void PartialSort<T, TComparer>(this Span<T> items, int middle, TComparer comparer)
        where TComparer : IComparer<T>;
    // Convenience overload
    public static void PartialSort<T>(this Span<T> items, int middle, Comparison<T> comparison);
}
```
`PartialSort` rearranges elements such that the range `[0, middle)` contains the sorted `middle` smallest elements in the range `[0, items.Length)`. The order of equal elements is not guaranteed to be preserved. The order of the remaining elements in the range `[middle, items.Length)` is unspecified. (modified from http://en.cppreference.com/w/cpp/algorithm/partial_sort)

This can be used with any type of memory, allows inlineable value type comparers. Note this would be restricted to `T` being `IComparable<T>` which is a bit different that my proposed `Sort` API, I want to revisit this for `Sort` too since this has some down sides... they should be identical in signature though for consistency.
